Codechange: Store animated tile state in map to improve performance.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

A game can become sluggish with lots (tens of thousands) of animated tiles.

Adding an animated tile involves searching a list to see if it's already present, before adding it.
Removing an animated tile again involves searching for list, and then also requires removing an element from anywhere in the vector.

If a game has many animated tile, the cost of this can be considerable.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Store animated tile state in map to improve performance.

This allows animated tiles to be added and removed without searching in the animated tile list, providing a performance improvement when there are lots of animated tiles.

Additionally, now that deleting animated tiles no longer happens immediately, they can be removed from the list by swapping with back of the list. Each delete is now a swap-and-pop operation which does not require removing from the middle of the list.

Map bits used by station tile flags are moved so that m6 bits 1..0 are available for all animated tiles.

Save game version is bumped so that animated tile state can be converted.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

We're still slower than JGRPP...

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

PeterN committed:
Add: Accessor to test if a tile may be animated.

PeterN committed:
Codechange: Move station tile flag cached bits from m6 to m3.

This frees up bits 0..1 of m6 for all tile types.

No savegame bump is needed as these bits are always reset on load.

glx22 commented on docs/landscape.html:
```diff
@@ -876,6 +877,9 @@ <h3><a name="Landscape">Landscape</a></h3>
+     <li>m3 bit 2: rail station / waypoint may have catenary pylons</li>
+     <li>m3 bit 1: rail station / waypoint may have catenary wires</li>
+     <li>m3 bit 0: rail station / waypoint is blocked</li>
      <li>m3 bits 3..2: ground type (road waypoints)
```
```suggestion
     <li>m3 bits 3..2: ground type (road waypoints)
     <li>m3 bit 2: rail station / waypoint may have catenary pylons</li>
     <li>m3 bit 1: rail station / waypoint may have catenary wires</li>
     <li>m3 bit 0: rail station / waypoint is blocked</li>
```
maybe for better 7 to 0 ordering.

glx22 reviewed (commented):
~Isn't station bits moving missing in `AfterLoad()`?~ It's fine, cached data always regenerated.

rubidium42 commented on docs/landscape_grid.html:
```diff
@@ -159,7 +159,7 @@ <h3 style="font-weight: bold;">Landscape</h3>
       <td class="bits" rowspan=2><span class="used" title="House type (m4 + m3[6])">XXXX XXXX</span></td>
       <td class="bits"><span class="used" title="Age in years, clamped at 255">XXXX XXXX</span></td>
-      <td class="bits" rowspan=2><span class="abuse" title="Newhouses activated: periodic processing time remaining; if not, lift position for houses 04 and 05">XXXX XX</span><span class="free">OO</span></td>
+      <td class="bits" rowspan=2><span class="abuse" title="Newhouses activated: periodic processing time remaining; if not, lift position for houses 04 and 05">XXXX XX</span><span class="used" title="Animated tile state">OO</span></td>
```
```suggestion
      <td class="bits" rowspan=2><span class="abuse" title="Newhouses activated: periodic processing time remaining; if not, lift position for houses 04 and 05">XXXX XX</span><span class="used" title="Animated tile state">XX</span></td>
```
It's used now, right?

rubidium42 commented on docs/landscape_grid.html:
```diff
@@ -185,10 +185,10 @@ <h3 style="font-weight: bold;">Landscape</h3>
       <td class="bits" rowspan=2><span class="used" title="Custom station specifications ID">XXXX XXXX</span></td>
       <td class="bits"><span class="used" title="Graphics index">XXXX XXXX</span></td>
-      <td class="bits" rowspan=2><span class="used" title="May have pylons">X</span><span class="used" title="Station type">XXXX</span> <span class="used" title="Reserved track">X</span><span class="used" title="May have wires">X</span><span class="used" title="Tile is blocked">X</span></td>
+      <td class="bits" rowspan=2><span class="free" title="May have pylons">X</span><span class="used" title="Station type">XXX X</span><span class="used" title="Reserved track">X</span><span class="used" title="Animated tile state">OO</span></td>
```
```suggestion
      <td class="bits" rowspan=2><span class="used" title="May have pylons">X</span><span class="used" title="Station type">XXX X</span><span class="used" title="Reserved track">X</span><span class="used" title="Animated tile state">XX</span></td>
```
Why has "may have pylons" become "free"? And if the tile state is used, why is it 00? I see no reason why it should remain 00.

rubidium42 commented on docs/landscape_grid.html:
```diff
@@ -313,7 +313,7 @@ <h3 style="font-weight: bold;">Landscape</h3>
       <td class="bits"><span class="free">OOOO OOOO</span></td>
       <td class="bits"><span class="pool" title="Object index on pool (m2 + m5)">XXXX XXXX</span></td>
-      <td class="bits"><span class="free">OOOO OOOO</span></td>
+      <td class="bits"><span class="free">OOOO OO</span><span class="used" title="Animated tile state">OO</span></td>
```
```suggestion
      <td class="bits"><span class="free">OOOO OO</span><span class="used" title="Animated tile state">XX</span></td>
```

rubidium42 commented on docs/landscape_grid.html:
```diff
@@ -280,7 +280,7 @@ <h3 style="font-weight: bold;">Landscape</h3>
       <td class="bits" rowspan=2><span class="used" title="Animation loop">XXXX XXXX</span></td>
       <td class="bits" rowspan=2><span class="used" title="Industry graphics ID (m5 + m6[2])">XXXX XXXX</span></td>
-      <td class="bits" rowspan=2><span class="free">OO</span><span class="used" title="Random triggers (NewGRF)">XXX</span> <span class="used" title="Industry graphics ID (m5 + m6[2])">X</span><span class="free">OO</span></td>
+      <td class="bits" rowspan=2><span class="free">OO</span><span class="used" title="Random triggers (NewGRF)">XXX</span> <span class="used" title="Industry graphics ID (m5 + m6[2])">X</span><span class="used" title="Animated tile state">OO</span></td>
```
```suggestion
      <td class="bits" rowspan=2><span class="free">OO</span><span class="used" title="Random triggers (NewGRF)">XXX</span> <span class="used" title="Industry graphics ID (m5 + m6[2])">X</span><span class="used" title="Animated tile state">XX</span></td>
```

PeterN committed:
Codechange: Store animated tile state in map to improve performance.

This allows animated tiles to be added and removed without searching in the animated tile list, providing a performance improvement when there are lots of animated tiles.

Save game version is bumped so that animated tile state can be converted.
